rcl_yaml_param_parser/src/node_params.c

```diff
@@ -19,25 +19,97 @@
   if (NULL == node_params->parameter_names) {
+    RCUTILS_SET_ERROR_MSG("Failed to allocate memory for node parameter names");
     return RCUTILS_RET_BAD_ALLOC;
   }
 
   node_params->parameter_values = allocator.zero_allocate(
-    MAX_NUM_PARAMS_PER_NODE, sizeof(rcl_variant_t), allocator.state);
+    capacity, sizeof(rcl_variant_t), allocator.state);
   if (NULL == node_params->parameter_values) {
+    allocator.deallocate(node_params->parameter_names, allocator.state);
+    node_params->parameter_names = NULL;
+    RCUTILS_SET_ERROR_MSG("Failed to allocate memory for node parameter values");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  node_params->num_params = 0U;
+  node_params->capacity_params = capacity;
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t node_params_reallocate(
+  rcl_node_params_t * node_params,
+  size_t new_capacity,
+  const rcutils_allocator_t allocator)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(node_params, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    &allocator, "invalid allocator", return RCUTILS_RET_INVALID_ARGUMENT);
+  // invalid if new_capacity is less than num_params
+  if (new_capacity < node_params->num_params) {
+    RCUTILS_SET_ERROR_MSG("invalid capacity");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  void * parameter_names = allocator.reallocate(
+    node_params->parameter_names, new_capacity * sizeof(char *), allocator.state);
+  if (NULL == parameter_names) {
+    RCUTILS_SET_ERROR_MSG("Failed to reallocate node parameter names");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+  node_params->parameter_names = parameter_names;
+  // zero initialization for the added memory
+  if (new_capacity > node_params->capacity_params) {
+    memset(
+      node_params->parameter_names + node_params->capacity_params, 0,
+      (new_capacity - node_params->capacity_params) * sizeof(char *));
+  }
+
+  void * parameter_values = allocator.reallocate(
+    node_params->parameter_values, new_capacity * sizeof(rcl_variant_t), allocator.state);
+  if (NULL == parameter_values) {
+    RCUTILS_SET_ERROR_MSG("Failed to reallocate node parameter values");
+    for (size_t parameter_idx = 0U; parameter_idx < node_params->num_params;
+      parameter_idx++)
+    {
+      char * param_name = node_params->parameter_names[parameter_idx];
+      if (NULL != param_name) {
+        allocator.deallocate(param_name, allocator.state);
+      }
+    }
     allocator.deallocate(node_params->parameter_names, allocator.state);
     node_params->parameter_names = NULL;
     return RCUTILS_RET_BAD_ALLOC;
   }
+  node_params->parameter_values = parameter_values;
+  // zero initialization for the added memory
+  if (new_capacity > node_params->capacity_params) {
+    memset(
+      &node_params->parameter_values[node_params->capacity_params], 0,
+      (new_capacity - node_params->capacity_params) * sizeof(rcl_variant_t));
+  }
 
+  node_params->capacity_params = new_capacity;
```

Comment:
I'll make one other suggestion about this whole function.  I think it might be nicer if we guarantee that on success, everything has been reallocated properly, but on failure, nothing has been changed in the structure.  Then it is simple for the caller of this function to determine what to do in either case.  To do that, I'll suggest something like the following pseudo-code:

```
void *names = allocator.reallocate(node_params->parameter_names, new_capacity);
if (!names) {
  return ERROR;
}
zero_new_names_memory();
void *values = allocator.reallocate(node_params->parameter_values, new_capacity);
if (!values) {
  allocator.deallocate(names);
  return ERROR;
}
zero_new_values_memory();
node_params->capacity_params = new_capacity;
node_params->parameter_names = names;
node_params->parameter_values = values;
return OK;
```

Also, please document this post-condition in the header file above.

The same basic idea applies to `rcl_yaml_node_struct_reallocate` below.

Reply:
> but on failure, nothing has been changed in the structure

Thanks for your suggestion.

But we can't promise that(Actually, I did that in my first commit.), because `reallocate` could return address values same as or different from the old address.

```
void *names = allocator.reallocate(node_params->parameter_names, new_capacity);    
if (!names) {
  return ERROR;
}
zero_new_names_memory();
void *values = allocator.reallocate(node_params->parameter_values, new_capacity);
if (!values) {
  allocator.deallocate(names);      // `step A`, please refer to below explanation.
  return ERROR;
}
```

At `step A`
- if `names` address is different from `node_params->parameter_names`, the content in `node_params->parameter_names` address is not exist (copy the memory to the new address and free it by reallocate). so I need to put `node_params->parameter_names = names;` before `reallocate values`. 
- use `allocator.deallocate(names); ` is dangerous(memory leak that old name items not be destroyed) if `names` is same as `node_params->parameter_names` after `reallocate node_params->parameter_names`
- if failure happened, we could keep unchanged by using two method
  - 1. `reallocate node_params->parameter_names ` with the original node_params->capacity_params. (But the address of node_params->parameter_names might be changed.)
  - 2. not to deallocate or reallocate `node_params->parameter_names`, just leave node_params->parameter_names as it was.

What do you think about that ?

Reply:
Thanks for the explanation; that makes sense.

In that case, I think we should just leave the code like it is; on failure, the structure will still contain all valid addresses.  The caller has to call `fini` on failure, but I think that is an OK constraint.